File: apps/workers/src/scalemyprints/api/schemas/spy.py
"""
Spy — request/response DTOs.

Mirror packages/contracts/src/spy.ts on the TypeScript side. Stays
deliberately thin: re-uses domain Pydantic models where possible.
"""

from __future__ import annotations

from datetime import datetime  # noqa: TC003
from typing import TYPE_CHECKING

from pydantic import BaseModel, ConfigDict, Field, HttpUrl

if TYPE_CHECKING:
    from scalemyprints.domain.spy.enums import (
        ImageMatchType,
        ListingStatus,
        Marketplace,
        VelocityClass,
    )
    from scalemyprints.domain.spy.models import (
        Listing,
        ReverseImageMatch,
        SpySearchResult,
    )

# ---------------------------------------------------------------------------
# POST /api/v1/spy/search
# ---------------------------------------------------------------------------


class SpySearchBody(BaseModel):
    """Body for POST /api/v1/spy/search."""

    model_config = ConfigDict(extra="forbid")

    text: str | None = Field(default=None, max_length=200)
    listing_url: HttpUrl | None = None
    marketplaces: list[Marketplace] = Field(default_factory=list)
    limit: int = Field(default=20, ge=1, le=100)


class SpyListingItem(BaseModel):
    """Public listing shape (one row in a search response)."""

    model_config = ConfigDict(extra="forbid")

    marketplace: Marketplace
    external_id: str
    url: HttpUrl
    title: str
    description: str | None
    tags: list[str]
    price_usd: float | None
    currency: str | None
    thumbnail_url: HttpUrl | None
    shop_handle: str | None
    shop_url: HttpUrl | None
    status: ListingStatus
    favorites: int | None
    reviews_count: int | None
    rating: float | None
    est_daily_sales: float | None
    velocity_class: VelocityClass
    first_seen_at: datetime
    last_seen_at: datetime

    @classmethod
    def from_domain(cls, listing: Listing) -> SpyListingItem:
        return cls(
            marketplace=listing.marketplace,
            external_id=listing.external_id,
            url=listing.url,
            title=listing.title,
            description=listing.description,
            tags=listing.tags,
            price_usd=listing.price_usd,
            currency=listing.currency,
            thumbnail_url=listing.thumbnail_url,
            shop_handle=listing.shop_handle,
            shop_url=listing.shop_url,
            status=listing.status,
            favorites=listing.favorites,
            reviews_count=listing.reviews_count,
            rating=listing.rating,
            est_daily_sales=listing.est_daily_sales,
            velocity_class=listing.velocity_class,
            first_seen_at=listing.first_seen_at,
            last_seen_at=listing.last_seen_at,
        )


class SpySourceFailure(BaseModel):
    """One adapter that errored during a fan-out search."""

    model_config = ConfigDict(extra="forbid")

    marketplace: Marketplace
    error: str


class SpySearchResponse(BaseModel):
    """Combined response for POST /api/v1/spy/search."""

    model_config = ConfigDict(extra="forbid")

    listings: list[SpyListingItem]
    sources_used: list[Marketplace]
    sources_failed: list[SpySourceFailure]
    total: int
    duration_ms: int

    @classmethod
    def from_domain(cls, result: SpySearchResult) -> SpySearchResponse:
        return cls(
            listings=[SpyListingItem.from_domain(l) for l in result.listings],
            sources_used=result.sources_used,
            sources_failed=[
                SpySourceFailure(marketplace=m, error=e) for m, e in result.sources_failed
            ],
            total=len(result.listings),
            duration_ms=result.duration_ms,
        )


# ---------------------------------------------------------------------------
# POST /api/v1/spy/reverse-image  (multipart/form-data)
# ---------------------------------------------------------------------------


class ReverseImageMatchItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    listing: SpyListingItem
    match_type: ImageMatchType
    phash_distance: int | None
    clip_cosine: float | None
    score: int

    @classmethod
    def from_domain(cls, match: ReverseImageMatch) -> ReverseImageMatchItem:
        return cls(
            listing=SpyListingItem.from_domain(match.listing),
            match_type=match.match_type,
            phash_distance=match.phash_distance,
            clip_cosine=match.clip_cosine,
            score=match.score,
        )


class ReverseImageResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    query_sha256: str
    matches: list[ReverseImageMatchItem]
    total: int
    duration_ms: int
    error: str | None = None


# ---------------------------------------------------------------------------
# GET /api/v1/spy/feed (hot movers)
# ---------------------------------------------------------------------------


class HotMoverItem(BaseModel):
    """A row from the `spy_hot_movers` view."""

    model_config = ConfigDict(extra="forbid")

    id: str
    marketplace: Marketplace
    title: str
    url: HttpUrl
    thumbnail_url: HttpUrl | None
    shop_handle: str | None
    shop_url: HttpUrl | None
    velocity_class: VelocityClass
    est_daily_sales: float | None
    price_usd: float | None
    favorites: int | None
    reviews_count: int | None
    last_seen_at: datetime


class HotMoversResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[HotMoverItem]
    total: int


# ---------------------------------------------------------------------------
# GET /api/v1/spy/shop/{marketplace}/{handle}
# ---------------------------------------------------------------------------


class ShopProfileItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    marketplace: Marketplace
    handle: str
    display_name: str | None
    url: HttpUrl
    location: str | None
    total_sales: int | None
    listings_count: int | None
    avg_review_rating: float | None
    reviews_count: int | None
    last_seen_at: datetime


# ---------------------------------------------------------------------------
# POST /api/v1/spy/shop-audit
# ---------------------------------------------------------------------------


class ShopAuditBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    marketplace: Marketplace
    handle: str = Field(min_length=1, max_length=120)
    depth: str = Field(default="standard", pattern="^(shallow|standard|deep)$")


class TagFrequencyItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    tag: str
    count: int = Field(ge=0)


class ShopAuditResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    shop: ShopProfileItem
    depth: str
    listings_sampled: int = Field(ge=0)
    est_monthly_revenue_usd: float | None
    avg_price_usd: float | None
    new_listings_last_30d: int | None
    restock_cadence_days: float | None
    top_listings: list[SpyListingItem]
    most_used_tags: list[TagFrequencyItem]
    captured_at: datetime
    error: str | None = None


# ---------------------------------------------------------------------------
# POST /api/v1/spy/saturation
# ---------------------------------------------------------------------------


class SaturationBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    phrase: str | None = Field(default=None, max_length=200)
    listing_ids: list[str] = Field(default_factory=list)
    marketplaces: list[Marketplace] = Field(default_factory=list)
    use_live_search: bool = True


class SaturationResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    score: int = Field(ge=0, le=100)
    saturation_class: str
    listings_count: int
    unique_shops: int
    hhi: float
    gmv_pool_usd: float
    density_component: int
    concentration_component: int
    velocity_component: int
    recency_component: int
    explanation: str


# ---------------------------------------------------------------------------
# POST /api/v1/spy/profit
# ---------------------------------------------------------------------------


class ProfitBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    marketplace: Marketplace
    product_type: str = Field(min_length=1, max_length=40)
    sale_price_usd: float = Field(gt=0)
    printer: str = Field(default="printify", min_length=1)
    shipping_usd: float = Field(default=0.0, ge=0.0)
    ad_cpc_usd: float = Field(default=0.0, ge=0.0)
    ad_conversion_rate: float = Field(default=0.0, ge=0.0, le=1.0)


class ProfitResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    sale_price_usd: float
    base_cost_usd: float
    marketplace_fee_usd: float
    shipping_usd: float
    ad_cost_usd: float
    profit_usd: float
    margin_pct: float
    printer: str
    note: str | None = None


# ---------------------------------------------------------------------------
# GET /api/v1/spy/ads
# ---------------------------------------------------------------------------


class AdSpyHitItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    platform: str
    ad_id: str
    page_or_handle: str
    page_id: str | None
    primary_text: str | None
    cta: str | None
    landing_url: HttpUrl | None
    started_at: datetime | None
    last_seen_at: datetime | None
    impressions_lower: int | None
    impressions_upper: int | None
    countries: list[str]


class AdLibraryResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    platform: str
    hits: list[AdSpyHitItem]
    total: int
    duration_ms: int
    error: str | None = None


# ---------------------------------------------------------------------------
# POST /api/v1/spy/_internal/refresh-velocity  (cron-only)
# ---------------------------------------------------------------------------


class VelocityRefreshBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    limit: int = Field(default=50, ge=1, le=500)
    max_age_hours: int = Field(default=6, ge=1, le=168)


class VelocityRefreshResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    started_at: datetime
    completed_at: datetime
    duration_ms: int
    candidates: int
    refreshed: int
    failed: int
    spikes_detected: int
    by_marketplace: dict[str, int]
    errors: list[str]
    reviews_count: int | None
    last_seen_at: datetime

File: packages/contracts/src/spy.ts
/**
 * Spy — shared contracts (TS).
 *
 * Mirrors apps/workers/src/scalemyprints/domain/spy/{enums,models}.py
 * and apps/workers/src/scalemyprints/api/schemas/spy.py.
 *
 * Keep these in lockstep — both sides use the same field names.
 */

export type Marketplace =
  | 'etsy'
  | 'amazon_merch'
  | 'redbubble'
  | 'teepublic'
  | 'society6'
  | 'zazzle'
  | 'spreadshirt'
  | 'bonanza'

export const MARKETPLACES: readonly Marketplace[] = [
  'etsy',
  'amazon_merch',
  'redbubble',
  'teepublic',
  'society6',
  'zazzle',
  'spreadshirt',
  'bonanza',
] as const

export const MARKETPLACE_LABELS: Record<Marketplace, string> = {
  etsy: 'Etsy',
  amazon_merch: 'Amazon Merch',
  redbubble: 'Redbubble',
  teepublic: 'Teepublic',
  society6: 'Society6',
  zazzle: 'Zazzle',
  spreadshirt: 'Spreadshirt',
  bonanza: 'Bonanza',
}

export type ListingStatus =
  | 'active'
  | 'sold_out'
  | 'inactive'
  | 'delisted'
  | 'unknown'

export type VelocityClass =
  | 'dormant'
  | 'steady'
  | 'rising'
  | 'spiking'
  | 'explosive'

export const VELOCITY_LABELS: Record<VelocityClass, string> = {
  dormant: 'Dormant',
  steady: 'Steady',
  rising: 'Rising',
  spiking: 'Spiking',
  explosive: 'Explosive',
}

export type SaturationClass = 'open' | 'mild' | 'crowded' | 'saturated'

export type ImageMatchType =
  | 'phash_exact'
  | 'phash_near'
  | 'clip_semantic'
  | 'clip_loose'

export const IMAGE_MATCH_LABELS: Record<ImageMatchType, string> = {
  phash_exact: 'Exact match',
  phash_near: 'Near-duplicate',
  clip_semantic: 'Same concept',
  clip_loose: 'Related design',
}

export type SpyFailureReason =
  | 'quota_exceeded'
  | 'source_blocked'
  | 'source_rate_limited'
  | 'invalid_input'
  | 'unsupported_marketplace'
  | 'image_too_large'
  | 'image_invalid'
  | 'embedding_failed'
  | 'storage_failure'
  | 'internal'

export type ViralSource =
  | 'reddit'
  | 'tiktok'
  | 'twitter'
  | 'instagram'
  | 'pinterest'
  | 'youtube_shorts'
  | 'google_trends'

// ---------------------------------------------------------------------------
// Core listing shape
// ---------------------------------------------------------------------------

export interface SpyListingItem {
  marketplace: Marketplace
  external_id: string
  url: string
  title: string
  description: string | null
  tags: string[]
  price_usd: number | null
  currency: string | null
  thumbnail_url: string | null
  shop_handle: string | null
  shop_url: string | null
  status: ListingStatus
  favorites: number | null
  reviews_count: number | null
  rating: number | null
  est_daily_sales: number | null
  velocity_class: VelocityClass
  first_seen_at: string
  last_seen_at: string
}

export interface SpySourceFailure {
  marketplace: Marketplace
  error: string
}

// ---------------------------------------------------------------------------
// POST /api/v1/spy/search
// ---------------------------------------------------------------------------

export interface SpySearchBody {
  text?: string | null
  listing_url?: string | null
  marketplaces?: Marketplace[]
  limit?: number
}

export interface SpySearchResponse {
  listings: SpyListingItem[]
  sources_used: Marketplace[]
  sources_failed: SpySourceFailure[]
  total: number
  duration_ms: number
}

// ---------------------------------------------------------------------------
// POST /api/v1/spy/reverse-image
// ---------------------------------------------------------------------------

export interface ReverseImageMatchItem {
  listing: SpyListingItem
  match_type: ImageMatchType
  phash_distance: number | null
  clip_cosine: number | null
  score: number
}

export interface ReverseImageResponse {
  query_sha256: string
  matches: ReverseImageMatchItem[]
  total: number
  duration_ms: number
  error: string | null
}

// ---------------------------------------------------------------------------
// GET /api/v1/spy/feed
// ---------------------------------------------------------------------------

export interface HotMoverItem {
  id: string
  marketplace: Marketplace
  title: string
  url: string
  thumbnail_url: string | null
  shop_handle: string | null
  shop_url: string | null
  velocity_class: VelocityClass
  est_daily_sales: number | null
  price_usd: number | null
  favorites: number | null
  reviews_count: number | null
  last_seen_at: string
}

export interface HotMoversResponse {
  items: HotMoverItem[]
  total: number
}

// ---------------------------------------------------------------------------
// Shop profile
// ---------------------------------------------------------------------------

export interface ShopProfileItem {
  marketplace: Marketplace
  handle: string
  display_name: string | null
  url: string
  location: string | null
  total_sales: number | null
  listings_count: number | null
  avg_review_rating: number | null
  reviews_count: number | null
  last_seen_at: string
}

// ---------------------------------------------------------------------------
// Shop audit
// ---------------------------------------------------------------------------

export interface TagFrequencyItem {
  tag: string
  count: number
}

export interface ShopAuditBody {
  marketplace: Marketplace
  handle: string
  depth?: 'shallow' | 'standard' | 'deep'
}

export interface ShopAuditResponse {
  shop: ShopProfileItem
  depth: string
  listings_sampled: number
  est_monthly_revenue_usd: number | null
  avg_price_usd: number | null
  new_listings_last_30d: number | null
  restock_cadence_days: number | null
  top_listings: SpyListingItem[]
  most_used_tags: TagFrequencyItem[]
  captured_at: string
  error: string | null
}

// ---------------------------------------------------------------------------
// Saturation / difficulty score
// ---------------------------------------------------------------------------

export interface SaturationBody {
  phrase?: string | null
  listing_ids?: string[]
  marketplaces?: Marketplace[]
  use_live_search?: boolean
}

export interface SaturationResponse {
  score: number              // 0-100
  saturation_class: SaturationClass
  listings_count: number
  unique_shops: number
  hhi: number
  gmv_pool_usd: number
  density_component: number
  concentration_component: number
  velocity_component: number
  recency_component: number
  explanation: string
}

// ---------------------------------------------------------------------------
// Profit calculator
// ---------------------------------------------------------------------------

export type ProductType =
  | 't_shirt'
  | 'tank_top'
  | 'long_sleeve'
  | 'hoodie'
  | 'sweatshirt'
  | 'mug_11oz'
  | 'mug_15oz'
  | 'tote_bag'
  | 'phone_case'
  | 'poster_18x24'
  | 'sticker'
  | 'blanket_50x60'
  | 'pillow_18x18'

export type PrinterId =
  | 'printful'
  | 'printify'
  | 'gelato'
  | 'customcat'
  | 'spod'

export const PRINTERS: readonly PrinterId[] = [
  'printify',
  'printful',
  'gelato',
  'customcat',
  'spod',
] as const

export const PRINTER_LABELS: Record<PrinterId, string> = {
  printify: 'Printify',
  printful: 'Printful',
  gelato: 'Gelato',
  customcat: 'CustomCat',
  spod: 'SPOD',
}

export const PRODUCT_LABELS: Record<ProductType, string> = {
  t_shirt: 'T-Shirt',
  tank_top: 'Tank Top',
  long_sleeve: 'Long Sleeve',
  hoodie: 'Hoodie',
  sweatshirt: 'Sweatshirt',
  mug_11oz: 'Mug (11oz)',
  mug_15oz: 'Mug (15oz)',
  tote_bag: 'Tote Bag',
  phone_case: 'Phone Case',
  poster_18x24: 'Poster (18×24)',
  sticker: 'Sticker',
  blanket_50x60: 'Blanket (50×60)',
  pillow_18x18: 'Pillow (18×18)',
}

export interface ProfitBody {
  marketplace: Marketplace
  product_type: ProductType
  sale_price_usd: number
  printer?: PrinterId
  shipping_usd?: number
  ad_cpc_usd?: number
  ad_conversion_rate?: number
}

export interface ProfitResponse {
  sale_price_usd: number
  base_cost_usd: number
  marketplace_fee_usd: number
  shipping_usd: number
  ad_cost_usd: number
  profit_usd: number
  margin_pct: number
  printer: PrinterId
  note: string | null
}

// ---------------------------------------------------------------------------
// Ad library
// ---------------------------------------------------------------------------

export type AdPlatform =
  | 'facebook'
  | 'instagram'
  | 'tiktok'
  | 'pinterest'
  | 'google_ads'

export interface AdSpyHitItem {
  platform: AdPlatform
  ad_id: string
  page_or_handle: string
  page_id: string | null
  primary_text: string | null
  cta: string | null
  landing_url: string | null
  started_at: string | null
  last_seen_at: string | null
  impressions_lower: number | null
  impressions_upper: number | null
  countries: string[]
}

export interface AdLibraryResponse {
  platform: AdPlatform
  hits: AdSpyHitItem[]
  total: number
  duration_ms: number
  error: string | null
}

// ---------------------------------------------------------------------------
// Velocity refresh (cron-only)
// ---------------------------------------------------------------------------

export interface VelocityRefreshResponse {
  started_at: string
  completed_at: string
  duration_ms: number
  candidates: number
  refreshed: number
  failed: number
  spikes_detected: number
  by_marketplace: Record<string, number>
  errors: string[]
}

// ---------------------------------------------------------------------------
// Phase 3 — viral mining, tag mining, TM overlay
// ---------------------------------------------------------------------------

export interface ViralSignalItem {
  source: ViralSource
  source_url: string | null
  phrase: string
  detected_at: string
  engagement: number
  momentum_score: number
  pod_readiness_score: number
  existing_pod_count: number
  suggested_styles: string[]
  note: string | null
}

export interface ViralFeedResponse {
  signals: ViralSignalItem[]
  sources_used: string[]
  sources_failed: { source: string; error: string }[]
  total: number
  duration_ms: number
}

export interface TagMineBody {
  seed: string
  marketplaces?: Marketplace[]
  per_marketplace_limit?: number
  top_n?: number
}

export interface MinedTagItem {
  tag: string
  total_count: number
  by_marketplace: Record<string, number>
  distinct_marketplaces: number
  sample_listings: string[]
}

export interface TagMineResponse {
  seed: string
  tags: MinedTagItem[]
  total_listings_scanned: number
  duration_ms: number
}

export interface TMOverlayBody {
  phrase: string
  marketplaces?: Marketplace[]
  nice_classes?: number[]
}

export type TMOverlayVerdict = 'go' | 'caution' | 'block'

export interface TMOverlayResponse {
  phrase: string
  opportunity_score: number
  risk_score: number
  saturation_score: number
  combined_verdict: TMOverlayVerdict
  listings_count: number
  est_monthly_gmv_usd: number
  trademark: {
    overall_risk_level: string
    overall_risk_score: number
    jurisdictions: {
      code: string
      risk_score: number
      risk_level: string
      match_count: number
      error: string | null
    }[]
    recommendations: {
      severity: string
      message: string
    }[]
  }
  duration_ms: number
}
